### PR TITLE
Add functions to generate preimage and preimage_hash

### PR DESCRIPTION
**Pull request by zhenlu**

### CODE REVIEW

File: src/signer.rs [hunk 1]
```diff
@@ -226,6 +227,25 @@ impl LightsparkSigner {
         Ok(self.build_commitment_secret(commitment_seed, per_commitment_point_idx))
     }
 
+    pub fn get_payment_preimage_nonce(&self) -> u64 {
```
Comment (jklein24) [inline src/signer.rs:230]: Will this be used anywhere? Isn't sparkcore generating the nonce?

**Review by jklein24 — APPROVED** [review]

File: src/signer.rs (hunk 1 above)
Comment (zhenlu) [reply src/signer.rs:230]: Client should generate the nonce, and then pass to sparkcore in the create_invoice mutation.

File: src/signer.rs [hunk 2]
```diff
@@ -3,11 +3,12 @@ use std::str::FromStr;
 use std::sync::Arc;
 
 use bitcoin::bip32::{DerivationPath, ExtendedPrivKey, ExtendedPubKey};
+use bitcoin::hashes::{HmacEngine, sha512, HashEngine, Hmac, Hash};
 use bitcoin::secp256k1::ecdh::SharedSecret;
 use bitcoin::secp256k1::ecdsa::Signature;
 use bitcoin::secp256k1::hashes::sha256;
 use bitcoin::secp256k1::{Message, PublicKey, Scalar, Secp256k1, SecretKey};
-use rand_core::OsRng;
+use rand_core::{OsRng, RngCore};
```
Comment (mhrheaume) [inline src/signer.rs:11]: I don't think `RngCore` is used?

File: src/signer.rs [hunk 3]
```diff
@@ -226,6 +227,25 @@ impl LightsparkSigner {
         Ok(self.build_commitment_secret(commitment_seed, per_commitment_point_idx))
     }
 
+    pub fn get_payment_preimage_nonce(&self) -> u64 {
+        let mut rng = OsRng;
+        rng.next_u64()
+    }
+
+    pub fn generate_preimage(&self, nonce: u64) -> Vec<u8> {
+        let key = self.derive_key("m/4h".to_owned()).unwrap();
```
Comment (mhrheaume) [inline src/signer.rs:238]: Does this ever return `Err`? If so, this should probably be a result rather than crash, right?

**Review by mhrheaume — APPROVED** [review]

File: src/signer.rs (hunk 3 above)
Comment (zhenlu) [reply src/signer.rs:238]: Ah good catch! I'll add to the return value.

File: src/signer.rs (hunk 2 above)
Comment (zhenlu) [reply src/signer.rs:11]: fill_bytes needs RngCore

**Pull request merged**